samples/nrf9160/mqtt_simple/src/main.c

```diff
@@ -0,0 +1,284 @@
+	struct addrinfo *addr;
+	struct addrinfo hints;
+
+	hints.ai_flags = 0;
+	hints.ai_family = AF_INET;
+	hints.ai_socktype = SOCK_STREAM;
+	hints.ai_protocol = 0;
+
+	err = getaddrinfo(CONFIG_MQTT_BROKER_HOSTNAME, NULL, &hints, &result);
+	if (err) {
+		printk("ERROR: getaddrinfo failed %d\n", err);
+
+		return;
+	}
+
+	addr = result;
+	err = -ENOENT;
+
+	/* Look for address of the broker. */
+	while (addr != NULL) {
+		/* IPv4 Address. */
+		if (addr->ai_addrlen == sizeof(struct sockaddr_in)) {
+			struct sockaddr_in *broker4 =
+				((struct sockaddr_in *)&broker);
+
+			broker4->sin_addr.s_addr =
+				((struct sockaddr_in *)addr->ai_addr)
+				->sin_addr.s_addr;
+			broker4->sin_family = AF_INET;
+			broker4->sin_port = htons(CONFIG_MQTT_BROKER_PORT);
+			printk("IPv4 Address found 0x%08x\n",
+				broker4->sin_addr.s_addr);
+			break;
+		} else {
+			printk("ai_addrlen = %d should be %d or %d\n",
+				addr->ai_addrlen,
+				sizeof(struct sockaddr_in),
+				sizeof(struct sockaddr_in6));
+		}
+
+		addr = addr->ai_next;
+		break;
+	}
+
+	/* Free the address. */
+	freeaddrinfo(result);
+}
+
+
+static void client_init(struct mqtt_client *client)
+{
+	mqtt_client_init(client);
+
+	broker_init();
+
+	/* MQTT client configuration */
+	client->broker = &broker;
+	client->evt_cb = mqtt_evt_handler;
+	client->client_id.utf8 = (u8_t *)CONFIG_MQTT_CLIENT_ID;
+	client->client_id.size = strlen(CONFIG_MQTT_CLIENT_ID);
+	client->password = NULL;
+	client->user_name = NULL;
+	client->protocol_version = MQTT_VERSION_3_1_1;
+
+	/* MQTT buffers configuration */
+	client->rx_buf = rx_buffer;
+	client->rx_buf_size = sizeof(rx_buffer);
+	client->tx_buf = tx_buffer;
+	client->tx_buf_size = sizeof(tx_buffer);
+
+	/* MQTT transport configuration */
+	client->transport.type = MQTT_TRANSPORT_NON_SECURE;
+}
+
+/**@brief Configures modem to provide LTE link. Blocks until link is
+ * successfully established.
+ */
+static void modem_configure(void)
+{
+	if (IS_ENABLED(CONFIG_LTE_AUTO_INIT_AND_CONNECT)) {
+		/* Do nothing, modem is already turned on
+		 * and connected.
+		 */
+	} else {
+		int err;
+
+		printk("LTE Link Connecting ...\n");
+		err = lte_lc_init_and_connect();
+		__ASSERT(err == 0, "LTE link could not be established.");
+		printk("LTE Link Connected!\n");
+	}
+}
+
+void main(void)
+{
+	int err;
+
+	printf("The MQTT simple sample started\n");
+
+	modem_configure();
+
+	if (IS_ENABLED(CONFIG_AT_HOST_LIBRARY)) {
+		err = at_host_init(CONFIG_AT_HOST_UART,
+			CONFIG_AT_HOST_TERMINATION);
+		if (err != 0) {
+			printf("ERROR: AT Host not initialized\n");
+			return;
+		}
+	}
+	if (IS_ENABLED(CONFIG_MQTT_LIB)) {
+		client_init(&client);
+
+		err = mqtt_connect(&client);
+		if (err != 0) {
+			printk("ERROR: mqtt_connect. %d\n", err);
+			return;
+		}
+	}
+
+	while (1) {
+		if (IS_ENABLED(CONFIG_AT_HOST_LIBRARY)) {
+			at_host_process();
+		}
+		if (IS_ENABLED(CONFIG_MQTT_LIB)) {
+			(void)mqtt_input(&client);
+			(void)mqtt_live(&client);
+		}
+		k_cpu_idle();
```

Comment:
I'm not sure if calling `k_cpu_idle` in a main thread is a good pattern to follow in examples. It's the idle thread that should do the idling, other threads should report themselves unready by using kernel primitives/functions, hence making the scheduler switch to the idle thread.

For MQTT this should be simpy achieved by calling `poll` function on MQTT socket, blocking the thread unless the data is ready. Not sure about the `at_host` though, whether it would be possible to block on it.

Reply:
@rlubos The at_host implementation must be changed first to not require polling at_host_process.
@jtguggedal Ok, so something like the asset_tracker main loop?

Reply:
@jarz-nordic  Do you have any input on what's the preferred way to make sure the logging thread gets time to process the logs when `main()` is busy looping?

Reply:
@nordic-krch is resposible for logging in Zephyr.